feat: diffusion baseline (including refactor) (#239)
p-doom/jasmine: README.md

        <img src="https://img.shields.io/badge/code%20style-black-000000.svg" /></a>
</p>

Jasmine is a production-ready JAX-based world modeling codebase. It currently implements the high-level architecture of [Genie: Generative Interactive Environments](https://arxiv.org/abs/2402.15391) (Bruce et al., 2024) with [MaskGIT](https://arxiv.org/abs/2202.04200) (Chang et al., 2022), a diffusion baseline, as well as an autoregressive (causal) baseline. 

Jasmine scales from single hosts to hundreds of xPUs thanks to XLA and strives to be an easily hackable, batteries-included foundation for world modeling research.

<h2 name="overview" id="overview">Overview</h2>

> Genie has three components: a [video tokenizer](jasmine/models/tokenizer.py), a [latent action model (LAM)](jasmine/models/lam.py), and a [dynamics model](jasmine/models/dynamics.py). 

Author: ```suggestion
Genie has three components: A [video tokenizer](jasmine/models/tokenizer.py), a [latent action model (LAM)](jasmine/models/lam.py), and a [dynamics model](jasmine/models/dynamics.py). 
```

> The tokenizer needs to be trained first, which the dynamics then uses. Jasmine supports co-training the LAM with the dynamics model.

Author: ```suggestion
The tokenizer needs to be trained first, which the dynamics subsequently uses. Jasmine supports co-training the LAM with the dynamics model.
```

> Jasmine implements a VQ-VAE and a MAE based tokenizer. The MaskGIT and causal baseline need to be trained using the discrete VQ-VAE tokenizer. The diffusion baseline is trained on latents from the MAE tokenizer.  The diffusion baseline, uses several elements from [Dreamer 4](https://arxiv.org/abs/2509.24527) (Hafner et. al., 2025) and is trained on the [diffusion forcing](https://arxiv.org/abs/2407.01392) (Chen et. al., 2024) objective.

Author: ```suggestion
The repository contains a VQ-VAE-based and a MAE-based tokenizer. The MaskGIT-based and causal baselines need to be trained using the discrete VQ-VAE tokenizer. The diffusion baseline uses the MAE tokenizer. The diffusion baseline adopts several (but not all) techniques from [Dreamer 4](https://arxiv.org/abs/2509.24527) (Hafner et. al., 2025) and uses the [diffusion forcing](https://arxiv.org/abs/2407.01392) (Chen et. al., 2024) objective for supervision.
```

<h2 name="train" id="train">Quick Start 🚀 </h2>

> For a quickstart you can train the tokenizer, latent action model and dynamics model sequentially as shown below. However, note that Jasmine supports various features, which can be configured through command line arguments. Therefore, please refer to the respective training scripts. 

Author: ```suggestion
```

I would remove this; we already mention the features in the readme.
